src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertificateProvider.cs

```diff
@@ -0,0 +1,149 @@
+    {
+        internal sealed class ClientCertificateProvider : IDisposable
+        {
+            internal readonly GCHandle _gcHandle;
+            internal readonly Interop.Ssl.ClientCertCallback _callback;
+            private SafeEvpPKeyHandle _privateKeyHandle = null;
+            private SafeX509Handle _certHandle = null;
+
+            internal ClientCertificateProvider ()
+            {
+                _gcHandle = GCHandle.Alloc(this);
+                _callback = TlsClientCertCallback;
+            }
+
+            private int TlsClientCertCallback(IntPtr ssl, out IntPtr certHandle, out IntPtr privateKeyHandle)
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(ssl);
+                using (SafeSslHandle sslHandle = new SafeSslHandle(ssl, false))
+                {
+                    certHandle = IntPtr.Zero;
+                    privateKeyHandle = IntPtr.Zero;
+                    VerboseTrace("libssl's client certificate callback");
+
+                    ISet<string> issuerNames = GetRequestCertificateAuthorities(sslHandle);
+                    X509Certificate2 certificate;
+                    X509Chain chain;
+                    if (!GetClientCertificate(issuerNames, out certificate, out chain))
+                    {
+                        VerboseTrace("no cert or chain");
+                        return 0;
+                    }
+
+                    Interop.Crypto.CheckValidOpenSslHandle(certificate.Handle);
+                    using (RSAOpenSsl rsa = certificate.GetRSAPrivateKey() as RSAOpenSsl)
+                    {
+                        if (rsa != null)
+                        {
+                            _privateKeyHandle = rsa.DuplicateKeyHandle();
+                        }
+                        else
+                        {
+                            using (ECDsaOpenSsl ecdsa = certificate.GetECDsaPrivateKey() as ECDsaOpenSsl)
+                            {
+                                if (ecdsa != null)
+                                {
+                                    _privateKeyHandle = ecdsa.DuplicateKeyHandle();
+                                }
+                            }
+                        }
+                    }
+
+                    if (_privateKeyHandle == null || _privateKeyHandle.IsInvalid)
+                    {
+                        VerboseTrace("invalid private key");
+                        return 0;
+                    }
+
+                    _certHandle = Interop.Crypto.X509Duplicate(certificate.Handle);
+                    Interop.Crypto.CheckValidOpenSslHandle(_certHandle);
+                    if (chain != null)
+                    {
+                        for (int i = chain.ChainElements.Count - 2; i > 0; i--)
+                        {
+                            SafeX509Handle dupCertHandle = Interop.Crypto.X509Duplicate(chain.ChainElements[i].Certificate.Handle);
```

Comment:
Fixed it but I reasoned that since if dupCertHandle is invalid, the immediate call would fail (```SslAddExtraChainCert```) would fail .